cupy/_core/_routines_linalg.pyx

```diff
@@ -449,7 +449,7 @@ cpdef ndarray tensordot_core(
         out = _ndarray_init(ret_shape, dtype)
     else:
         if out.dtype != dtype:
-            out = _ndarray_init(ret_shape, dtype)
+            raise NotImplementedError("The out array dtype is mismatched")
```

Comment:
I agree that the current implementation, which ignores `out`, seems wrong.  I'm not sure if the change is better than commenting `# TODO: Fix to write to out`.

Reply:
FWIW NumPy's `matmul` will just error in this case.

```python
import numpy as np

a = np.random.random((2, 3))
b = np.random.random((3, 2))
c = np.empty((2, 2), dtype=int)

np.matmul(a, b, out=c)
```

```python
---------------------------------------------------------------------------
UFuncTypeError                            Traceback (most recent call last)
<ipython-input-7-a4f34170f335> in <module>
      5 c = np.empty((2, 2), dtype=int)
      6 
----> 7 np.matmul(a, b, out=c)

UFuncTypeError: Cannot cast ufunc 'matmul' output from dtype('float64') to dtype('int64') with casting rule 'same_kind'
```

<br>

Note: `UFuncTypeError ` is [just a `TypeError` subclass]( https://github.com/numpy/numpy/blob/f4353323d0be0c3b3c7d0c752d6c135bc6d36465/numpy/core/_exceptions.py#L32 )